pkg/api/query/v1.go

```diff
@@ -540,21 +540,30 @@ func (qapi *QueryAPI) labelValues(r *http.Request) (interface{}, []error, *api.A
 		vals     []string
 		warnings storage.Warnings
 	)
-	// TODO(yeya24): push down matchers to Store level.
 	if len(matcherSets) > 0 {
-		// Get all series which match matchers.
-		var sets []storage.SeriesSet
-		for _, mset := range matcherSets {
-			s := q.Select(false, nil, mset...)
-			sets = append(sets, s)
+		var callWarnings storage.Warnings
+		labelValuesSet := make(map[string]struct{})
```

Comment:
My only question is why are you building a `map[string]struct{}` here instead of adding everything directly to vals:

```go
			retrievedVals, callWarnings, err := q.LabelValues(name, matchers...)
			if err != nil {
				return nil, nil, &api.ApiError{Typ: api.ErrorExec, Err: err}
			}
			warnings = append(warnings, callWarnings...)
			vals = append(vals, retrievedVals...)
```
? :thinking: This should be more efficient.

Reply:
I think this is to filter out duplicates.

Reply:
Ahh, right! Thanks for the answer.